### ISSUE TITLE
Full compatibility with Python 3.4 import spec

### ISSUE THREAD

**Issue by olemis** [brython-dev/brython#189]
Incompatibilities detected in Brython import machinery

  * Import hooks not implemented ?

```python

Brython 3.1.3 on Netscape 5.0 (Macintosh; Intel Mac OS X 10_9_2)
 AppleWebKit/537.36 (KHTML, like Gecko) Chrome/40.0.2214.91
 Safari/537.36
 >>> import sys
 >>> sys.path
 ['http://localhost/brython/www/src/Lib',
 'http://localhost/brython/www/tests',
 'http://localhost/brython/www/src/Lib/site-packages']
 >>> sys.meta_path
 []
 >>> def finder_factory(nm, path):
 ...     class F:
 ...         def find_module(self):
 ...             print('Ok')
 ...             return None
 ...     return F()
 ...
 >>> sys.meta_path.append(finder_factory)
 >>> sys.meta_path
 [<function finder_factory>]
 >>> import whatever
 Traceback (most recent call last):
   module __main__17 line 116
     _ = editor_ns['_'] = eval(currentLine, editor_ns)
   module __main__17 line 116
     _ = editor_ns['_'] = eval(currentLine, editor_ns)
   module __main__17 line 116
     _ = editor_ns['_'] = eval(currentLine, editor_ns)
   module __main__17 line 116
     _ = editor_ns['_'] = eval(currentLine, editor_ns)
   module __main__17 line 118
     print(repr(_))
   module __main__17 line 116
     _ = editor_ns['_'] = eval(currentLine, editor_ns)
   module __main__17 line 131
     exec(currentLine, editor_ns)
   module exec_64 line 1
     import whatever
 ImportError: cannot import whatever
 >>>
```

  * ImportError not raised if module cache entry is set to None

```python

>>> sys.modules['x'] = None
>>> import x
>>> repr(x)
'None'
```

**Comment by olemis** [comment]
I'll be approaching the upgrade of Brython import machinery in phases

* Right now I'm baking a patch for #189 aimed at getting closer
  to specs which includes
  * A ready-to-use sys module at Brython boot time [3]_
  * ... initialized with every current import strategy implemented
    as a meta path hook
  * Support for hook methods exec_module et al. [1]_
  * __import__ overrides [1]_
- Async imports , see #188
  * This one looks challenging 
- Implement importlib , see #192
* Advanced import features
  * PEP 420 [6]_ 
  * PEP 328 [7]_ 
  * PEP 366 [8]_ 
  * PEP 451 [9]_ 
  * PEP 3120 [10]_ 
  * PEP 3147 [11]_ 
* Wrap-up for full compatibility and testing

.. [1] https://docs.python.org/3/reference/import.html

.. [2] https://www.python.org/dev/peps/pep-0302/

.. [3] https://docs.python.org/3/reference/toplevel_components.html#programs

.. [4] https://docs.python.org/3/reference/simple_stmts.html#import

.. [5] https://docs.python.org/3/library/functions.html#__import__

.. [6] https://www.python.org/dev/peps/pep-0420/

.. [7] https://www.python.org/dev/peps/pep-0328/

.. [8] https://www.python.org/dev/peps/pep-0366/

.. [9] https://www.python.org/dev/peps/pep-0451/

.. [10] https://www.python.org/dev/peps/pep-3120/

.. [11] https://www.python.org/dev/peps/pep-3147/

**Comment by PierreQuentel** [comment]
This was fixed in commit 4df83b8b5540fdcf4a37180feea9d18367520b21 with code by Olemis

**Comment by olemis** [comment]
@PierreQuentel I reviewed your work and there is at least one thing that is not completely accurate. In `new_spec` function the object returned back to the caller is a module object .

Instead it is necessary to add stdlib [https://docs.python.org/3/library/importlib.html#importlib.machinery.ModuleSpec](ModuleSpec type) and bind it to this instance's `__class__` attribute . This is documented in [http://www.python.org/dev/peps/pep-0451](PEP 451) . I've created #233 for this purpose .
